Customization args fixes

nikitamarysolomanpvt opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
We were facing issue in retrieving data from customization_args for multiple-choice, where we were expecting StringList. 
issue found:
objectType_ = OBJECTTYPE_NOT_SET

Found `is List<*>` expected in `createCustomizationArgValueFromJson` where as we were having JSONArray.
It can be solved by using Gson and converting it to ArrayList.

**We may iterate and create an ArrayList of String from JSONArray to avoid usage of Gson**
expected result retrieved:
by using setOfHtmlString.htmlList
```
result = {ProtobufArrayList@12654}  size = 3
 0 = "It's translated from a different language."
 1 = "It's a nonsense word that someone made up."
 2 = "It's the name of a popular cartoon character."
```

nikitamarysolomanpvt committed:
customization_args issues fix

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

rt4914 commented:
@nikitamarysolomanpvt  If this PR is ready for review, please assign it to concerned developers.

nikitamarysolomanpvt committed:
nit

veena14cs commented on domain/build.gradle:
```diff
@@ -43,6 +43,7 @@ dependencies {
       'androidx.appcompat:appcompat:1.0.2',
       'androidx.lifecycle:lifecycle-livedata-ktx:2.2.0-alpha03',
       'com.google.dagger:dagger:2.24',
+      'com.google.code.gson:gson:2.8.6',
```
make it alphabetical order.

nikitamarysolomanpvt replied:
Resolved

nikitamarysolomanpvt committed:
nit

veena14cs commented:
LGTM

rt4914 commented:
@nikitamarysolomanpvt I think i am not the right person to review this PR. Though it is functional but in such cases approach can matter a lot. So I think @BenHenning  should review this PR and I will unassign myself.

BenHenning commented on domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt:
```diff
@@ -272,10 +273,14 @@ class ExplorationRetriever @Inject constructor(private val context: Context) {
-        ).build()
+      else -> {
+        var customizationArgValueTemp: ArrayList<*> =
+          Gson().fromJson(customizationArgValue.toString(), ArrayList::class.java)
```
Can you briefly explain in  a comment why we need to rely on Gson here and why we can't parse the list using Android's JSON parser?

I'm not sure myself why we need to do this approach. What is the type of customizationArgValue in the set of HTML case?

BenHenning replied:
Also by 'type' here I mean what the actual object type is. I see in your PR base comment that the reported type of the value is OBJECTTYPE_NOT_SET, but I'm guessing that means the object type is JSONObject. Please let me know if that's not the case.

nikitamarysolomanpvt replied:
@BenHenning 
Yes object type is JSONArray .
Code is updated with out using Gson by iterating and adding to a MutableList

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -121,7 +121,7 @@ class StateFragmentPresenter @Inject constructor(
   private fun subscribeToCurrentState() {
     ephemeralStateLiveData.observe(fragment, Observer<EphemeralState> { result ->
-      logger.d("StateFragment", "getCurrentState: ${result.state.name}")
+      logger.d("StateFragment", "getCurrentState: ${result.state.interaction.customizationArgsMap.get("choices")!!.setOfHtmlString.htmlList}")
```
I suspect this was here for debug purposes, but it's a misleading log statement to check in. Please revert.

nikitamarysolomanpvt replied:
Reverted this change.

BenHenning reviewed (changes requested):
Thanks @nikitamarysolomanpvt! I have one question about why GSON is needed. PTAL.

rt4914 commented:
@nikitamarysolomanpvt What is the status of this PR?

Either finish this PR or close this PR.

nikitamarysolomanpvt committed:
createCustomizationArgValueFromJson type JsonArry as type to get StringList for htmlString.
updated createStringList().

Merge branches 'customization-args-fixes' and 'develop' of https://github.com/oppia/oppia-android into customization-args-fixes

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
#	domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt

nikitamarysolomanpvt committed:
createCustomizationArgValueFromJson type JsonArry as type to get StringList for htmlString.
updated createStringList().

Merge branches 'customization-args-fixes' and 'develop' of https://github.com/oppia/oppia-android into customization-args-fixes

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
#	domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt

nikitamarysolomanpvt committed:
createCustomizationArgValueFromJson type JsonArry as type to get StringList for htmlString.
updated createStringList().

Merge branches 'customization-args-fixes' and 'develop' of https://github.com/oppia/oppia-android into customization-args-fixes

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt
#	domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

BenHenning commented on app/src/main/res/layout/home_fragment.xml:
```diff
@@ -30,7 +30,7 @@
         android:onClick="@{presenter::playExplorationButton}"
         android:text="Play exploration"
-        android:visibility="gone" />
+        android:visibility="visible" />
```
Please hide this test only button.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on domain/src/main/java/org/oppia/domain/util/StateRetriever.kt:
```diff
@@ -287,26 +289,21 @@ class StateRetriever @Inject constructor(
-    val stringList = mutableListOf<String>()
-    if (value[0] is String) {
-      stringList.addAll(value as List<String>)
+  private fun createStringList(stringList: MutableList<String>): StringList {
```
I think this can be just a ``List``.

BenHenning replied:
Also, could we just take a JSONArray here, instead, to simplify?

nikitamarysolomanpvt replied:
Done

BenHenning commented on utility/src/main/java/org/oppia/util/parser/JsonParser.kt:
```diff
@@ -0,0 +1,14 @@
+
+/** This class parse json to mutable list of String. */
+class JsonParser {
+  fun parseJsonArrayToMutableListOfString(jsonArray: JSONArray): MutableList<String> {
```
Let’s keep this in the other file rather than introducing a new class and method. 

Also, this can just return ``List``.

Finally, we can probably shorten the method name: parseJsonStringList.

nikitamarysolomanpvt replied:
Resolved

BenHenning commented on utility/src/main/java/org/oppia/util/parser/JsonParser.kt:
```diff
@@ -0,0 +1,14 @@
+class JsonParser {
+  fun parseJsonArrayToMutableListOfString(jsonArray: JSONArray): MutableList<String> {
+    val list: MutableList<String> = ArrayList()
+    for (i in 0..jsonArray!!.length() - 1) {
```
Why is !! needed here? The jsonArray variable is non-null.

nikitamarysolomanpvt replied:
Resolved

BenHenning reviewed (approved):
Thanks @nikitamarysolomanpvt. Please let me know if you have any questions regarding my comments, otherwise fee free to merge once they are resolved.

nikitamarysolomanpvt committed:
Removed JsonParser and added the funcationality in StateRetriver

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

rt4914 reviewed (approved):
LGTM
